### ISSUE TITLE
Debug slow uploads

### ISSUE THREAD

**Issue by CyberCyclone** [humanmade/S3-Uploads#466]
Since using this plugin, our image uploads have gone from about 3 seconds per image to about 40 seconds. From our New Relic tracing, it appears that the image resizing is taking the most time. From what I can understand from the code, the S3-Uploads plugin is extending the default WordPress WP_Image_Editor_Imagick class and simply rewriting the functions to load them to the server, resize, and then write back to S3. Is this correct?

I'm guessing that if this is the case, it's saving the original file to the uploads folder (in the S3 bucket), then pulling that from the bucket for each image resize. As, is it pulling the original image for each specific resize? So if you have an original image that's 10MB, does it download the original 10MB all over again for each resize?

Our servers are in Australia, and we're using Digital Ocean Spaces in San Francisco (it's the closest available to Australia with about 300ms response), so there's going to be network bottleneck. We're trying to work out if 40 seconds too process an image is reasonably typical, or are people far better results than this? Our typical original image size is about 250KB to 1MB.

Thanks!

![image](https://user-images.githubusercontent.com/4373888/98197436-6ff51e80-1f72-11eb-83ef-7fd256b1730e.png)

**Comment by CyberCyclone** [comment]
If the above is true, would it be an option keep a copy of the original file on the server for about 10 minutes and use that as a reference? I don't know if that would be an issue if the same file name is uploaded again as you would need to know somehow that it's a different file and then delete it once it's no longer needed.

**Comment by hfoletto** [comment]
Did you manage to fix this issue?

**Comment by CyberCyclone** [comment]
No, it's still an issue. The most we could do is try and use a S3 bucket in a closer region. 

The main issue with the approach that this plugin has taken is that reading the file so many times adds to the outbound / retrieval costs.

**Comment by luco** [comment]
Having this issue as well. Without using this plugin, our API creates a woocommerce product in 20s. With this plugin, 1m:30s.

**Comment by rmccue** [comment]
Sorry for the delay in getting to this. This sounds like an issue with the S3-compatible service you're using. We don't see times anywhere like this using S3.

For example, I just checked a live server running in eu-west-1, and uploading a 500KB file spent 72ms in the PUT call to S3.

If you use S3 in ap-southeast-2 (Sydney), do you see the same response times?

> If the above is true, would it be an option keep a copy of the original file on the server for about 10 minutes and use that as a reference?

No; this wouldn't work in multi-server architectures, which is what S3 Uploads is primarily designed for.

**Comment by carstenwawer** [comment]
We are experiencing the same problems here. Calling WP_Image_Editor_Imagick :: thumbnail_image regularly takes over 40 seconds for a relatively small image.

Has anyone found a solution?

**Comment by carstenwawer** [comment]
I was able to solve the problem by switching from Imagemagick to GD. I'm still unsure why Imagemagick has such problems in combination with s3-uploads, but for me the performance has reduced to the normal level again.

`function szs_set_default_image_library( $editors ) {
	$gd_editor = 'WP_Image_Editor_GD';
	$editors = array_diff( $editors, array( $gd_editor ) );
	array_unshift( $editors, $gd_editor );
	return $editors;
}
add_filter( 'wp_image_editors', 'szs_set_default_image_library' );`

Does that help you, @CyberCyclone ?

**Comment by CyberCyclone** [comment]
> I was able to solve the problem by switching from Imagemagick to GD. I'm still unsure why Imagemagick has such problems in combination with s3-uploads, but for me the performance has reduced to the normal level again.
> 
> 
> 
> > function szs_set_default_image_library( $editors ) {
> 
> > 	$gd_editor = 'WP_Image_Editor_GD';
> 
> > 	$editors = array_diff( $editors, array( $gd_editor ) );
> 
> > 	array_unshift( $editors, $gd_editor );
> 
> > 	return $editors;
> 
> > }
> 
> > add_filter( 'wp_image_editors', 'szs_set_default_image_library' );
> 
> 
> 
> Does that help you, @CyberCyclone ?

We've since moved our S3 location and main servers closer together. We do still see high network traffic with constant retrieving of the original image.

**Comment by VlasicVojislav** [comment]
Closing this as an old issue.
